### PR TITLE
Lute/reification step4 testing ast

### PR DESCRIPTION
**Pull request by lute47lillo**
SymbolicTerm to Function Transformer.

Tests still needed to verify all working.

### CODE REVIEW

**Comment by lute47lillo** [comment]
@jorgefandinno Okay, I've found some problems with how we are parsing strong negated and negated atoms when using reification. I just realized the issue of using "-" + name on the reify strong negation. Now, the expected program is an UnaryOperator, but the parsed program returns a function of name "-a" which I believe is incorrect. 

I believe that I'll have to make some changes on top of the previous reification changes to do it correctly, and to test for equal programs instead of strings.

File: src/eclingo/parsing/transformers/parser_negations.py [hunk 1]
```diff
@@ -44,12 +45,15 @@ def visit_UnaryOperation(self, x):
...
 
         if self.reification:
-            aux_name = "-" + name
+            aux_name = name
         else:
             aux_name = self.strong_negation_prefix + "_" + name
             self.replacement.add((name, len(arguments), aux_name))
 
         atom = ast.Function(location, aux_name, arguments, external)
+        if self.reification:
```
Comment (lute47lillo) [inline src/eclingo/parsing/transformers/parser_negations.py:54]: I've finally solved that issue when using reification. The problem was the declaration of the name "-a" of the Function when it should have been an UnaryOperation that had an ast.Function as argument with the name "a".

File: src/eclingo/parsing/transformers/parser_negations.py [hunk 2]
```diff
@@ -44,12 +45,15 @@ def visit_UnaryOperation(self, x):
...
+            return ast.UnaryOperation(location, 0, atom)
+
```
Comment (lute47lillo) [inline src/eclingo/parsing/transformers/parser_negations.py:56]: I've finally solved that issue when using reification. The problem was the declaration of the name "-a" of the Function when it should have been an UnaryOperation that had an ast.Function as argument with the name "a".

File: tests/test_reification.py [hunk 3]
```diff
@@ -33,118 +31,142 @@ def parse_program(stm, parameters=[], name="base"):
...
-        self.assertListEqual(sorted(program), sorted(expected_program))
-
+        expected_program = function_transformer._rule_to_symbolic_term_adapter(expected_program)
+        
+        #print("\nThe expected program\n", expected_program)
+        #print("\nThe parsed program\n", program)
+        
+        
+        sorted_program = sorted(program)
+        expected_program.sort()
```
Comment (lute47lillo) [inline tests/test_reification.py:53]: But, now I'm trying to figure out some issues for the asserting case of the lists being equal. While they are equal, they are being ordered in the wrong order and that is causing tests to fail. I'll look into it tomorrow

File: tests/test_reification.py [hunk 4]
```diff
@@ -33,118 +31,142 @@ def parse_program(stm, parameters=[], name="base"):
...
             parse_program(":- &k{- -a}."), ":- k(u(a)). {k(u(a))} :- u(a)."
         )
+    '''
     
+    # Fail
     def test_epistemic_atom_with_default_negation(self):
         self.assert_equal_program(
             parse_program(":- &k{ not a}."),
             ":- k(not1(u(a))). not1(u(a)) :- not u(a). {k(not1(u(a)))} :- not1(u(a)).",
         )
```
Comment (lute47lillo) [inline tests/test_reification.py:113]: Specially, I've seen it happening here. The tests with a #Fail above are the ones facing this issue. The rest of the tests are passing.

Comment (jorgefandinno) [reply tests/test_reification.py:113]: How are each of them ordered?

Comment (lute47lillo) [reply tests/test_reification.py:113]: There you can see how the sorted program (which is the reference correctly parsed program) first has the rule not1(u(a)) :- ... while the expected program (The one we are parsing from the expected string) has the conditional Literal first

**Comment by lute47lillo** [comment]
The first part of the tests that were failing due to the bug are now failing.

However, I'm having some issues with these ones:
https://github.com/jorgefandinno/eclingo/blob/facbdc1c6f5c31be6c53ac15c900d568230160d9/tests/test_reification.py#L215-L238

I'll check them later, but it may have something to do with the newly introduced transformer.

Pylint is also failing, but I do not see where the error is.

**Comment by lute47lillo** [comment]
> For the tests:
> 
> def test_parameters01(self): self.assert_equal_program( parse_program("a(1..n).", ["n"], "parametrized"), "#program parametrized(n). u(a(1..n)).", )
> 
> ```
>  def test_parameters02(self): 
>      self.assert_equal_program( 
>          parse_program("a(1..n).", ["n"], "base"), "#program base(n). u(a(1..n))." 
>      ) 
> ```
> 
> I'm getting another error, which the sorted program evaluates the Interval 1..n, as being Symbolic term of number 1 and another Symbolic term of a clingo Function with name 'n'.
> 
> However, the way the Transformer is set, is that the 2nd Symbolic term is being transformed to an AST.Function of name 'n' and therefore failing the test.

The idea I've had is to create another visit_Interval after this Transformer is being parsed and re-do the conversion of the interval components to Symbolic Terms,

would that work?

**Pull request merged**